File: 2020/11/both.pl
#!/usr/bin/perl

use strict;
use integer;

my $fh;

die "* no input\n" if (!open($fh, '<', $ARGV[0] || 'input.txt'));

my $linenum = 0;

my $map;
my $stride;
my $rows;
while (1)
{
	my $line = <$fh>;
	$linenum++;
	last if (!$line);
	chomp($line);
	$map .= " $line";
	$rows++;
	$stride = length($map) if (!$stride);
}

my $blank = " " x $stride;
$map = $blank . $map . $blank . ' ';

my $map2 = $map;

while (1) {
	my $newmap = generation($map);
	last if ($newmap eq $map);
	$map = $newmap;
}

my $answer1 = $map =~ tr/#//;

printf("Part 1 answer: %d\n", $answer1);

while (1) {
	my $newmap = generation2($map2);
	last if ($newmap eq $map2);
	$map2 = $newmap;
}

my $answer2 = $map2 =~ tr/#//;

printf("Part 2 answer: %d\n", $answer2);

sub generation
{
	my ($map) = @_;
	my $newmap = $map;
	my $end = length($map) - $stride - 1;

	for (my $pos = $stride + 1; $pos <= $end; $pos++)
	{
		my $near = substr($map, $pos - $stride - 1, 3) .
			   substr($map, $pos - 1, 3) .
			   substr($map, $pos + $stride - 1, 3);
		if ((substr($near, 4, 1) eq 'L') && ($near !~ /#/))
		{
			substr($newmap, $pos, 1) = '#';
		}
		elsif ((substr($near, 4, 1) eq '#') && ($near =~ tr/#// >= 5))
		{
			substr($newmap, $pos, 1) = 'L';
		}
	}
	return $newmap;
}

sub generation2
{
	my ($map) = @_;
	my $newmap = $map;
	my $end = length($map) - $stride - 1;
	my @dirs =
	(
		-$stride - 1,
		-$stride,
		-$stride + 1,
		-1,
		1,
		$stride - 1,
		$stride,
		$stride + 1,
	);

	for (my $pos = $stride + 1; $pos <= $end; $pos++)
	{
		my $count = 0;
		my $atpos;
		foreach my $dir (@dirs)
		{
			my $scanpos = $pos;
			do
			{
				$scanpos += $dir;
				$atpos = substr($map, $scanpos, 1);
			} while ($atpos eq '.');
			$count++ if ($atpos eq '#');
		}
		$atpos = substr($map, $pos, 1);
		if (($atpos eq 'L') && ($count == 0))
		{
			substr($newmap, $pos, 1) = '#';
		}
		elsif (($atpos eq '#') && ($count >= 5))
		{
			substr($newmap, $pos, 1) = 'L';
		}
	}
	return $newmap;
}

sub print_map
{
	my ($map) = @_;

	print(join("\n", split(/ +/, $map)), "\n");
}

File: 2020/11/both.py
#!/usr/bin/python

import sys
import re

infile = 'input.txt'
if (len(sys.argv) > 1):
	infile = sys.argv[1]

fh = open(infile, 'r')

stride = 0
rows = 0
map = ''

for line in fh:
	map += ' ' + line.rstrip()
	rows += 1
	if not stride:
		stride = len(map)

blank = ' ' * stride
map = blank + map + blank + ' '
map2 = map

def print_map(map):
	for x in map.split():
		print(x)
	print

def generation(map):
	newmap = ""
	for pos in range(len(map)):
		c = map[pos]
		near = map[(pos - stride - 1) : (pos - stride + 2)]
		near += map[(pos - 1) : (pos + 2)]
		near += map[pos + stride - 1: pos + stride + 2]
		if c == 'L' and near.count('#') == 0:
			c = '#'
		elif c == '#' and near.count('#') >= 5:
			c = 'L'
		newmap += c

	return newmap

def generation2(map):
	newmap = ""
	dirs = (-stride - 1, -stride, -stride + 1,
	        -1, 1,
	        stride - 1, stride, stride + 1)

	for pos in range(len(map)): 
 		count = 0
 		atpos = 0
 		c = map[pos]
 		if c != ' ':
	 		for dir in dirs:
 				scanpos = pos
 				while True:
 					scanpos += dir
 					atpos = map[scanpos]
 					if atpos != '.':
 						break
 				if atpos == '#':
 					count += 1
 		if c == 'L' and count == 0:
 			c = '#'
 		elif c == '#' and count >= 5:
 			c = 'L'
		newmap += c

	return newmap

while True:
	newmap = generation(map)
	if newmap == map:
		break
	map = newmap

answer1 = map.count('#')
print("Part 1 answer: %d" % answer1)

while True:
	newmap = generation2(map2)
	if newmap == map2:
		break
	map2 = newmap

answer2 = map2.count('#')

print("Part 2 answer: %d" % answer2)
